File: KeySAVCoreJS/PKX.cs
using System;
using JS;

namespace KeySAVCore.Structures
{
    public struct PKX
    {
        public readonly uint 
            ec, pid, exp, 
            evHp, evAtk, evDef, evSpAtk, evSpDef, evSpe,
            ivHp, ivAtk, ivDef, ivSpe, ivSpAtk, ivSpDef,
            contestStatCool, contestStatBeauty, contestStatCute, contestStatSmart, contestStatTough, contestStatSheen,
            markings, hpType;

        public readonly string
            nickname, notOT, ot;

        public readonly int
            pkrsStrain, pkrsDuration,
            levelMet, otGender;

        public readonly bool
            isEgg, isNick, isShiny, isGhost, isFatefulEncounter;

        public readonly ushort
            ability, abilityNum, nature,
            species, heldItem, tid, sid, tsv, esv,
            move1, move2, move3, move4,
            move1Pp, move2Pp, move3Pp, move4Pp,
            move1Ppu, move2Ppu, move3Ppu, move4Ppu,
            eggMove1, eggMove2, eggMove3, eggMove4,
            chk,

            otFriendship, otAffection,
            eggLocation, metLocation,
            ball, encounterType,
            gamevers, countryID, regionID, dsregID, otLang;

        public readonly short 
            box, slot;

        public readonly byte
            form, gender;

        public readonly Date
            metDate, eggDate;

        public PKX(Uint8Array pkx, short box, short slot, bool isghost)
        {
            this.box = box;
            this.slot = slot;

            this.isGhost = isghost;

            nickname = "";
            notOT = "";
            ot = "";
            ec = BitConverter.ToUInt32(pkx, 0);
            chk = BitConverter.ToUInt16(pkx, 6);
            species = BitConverter.ToUInt16(pkx, 0x08);
            heldItem = BitConverter.ToUInt16(pkx, 0x0A);
            tid = BitConverter.ToUInt16(pkx, 0x0C);
            sid = BitConverter.ToUInt16(pkx, 0x0E);
            exp = BitConverter.ToUInt32(pkx, 0x10);
            ability = pkx[0x14];
            abilityNum = pkx[0x15];
            // 0x16, 0x17 - unknown
            pid = BitConverter.ToUInt32(pkx, 0x18);
            nature = pkx[0x1C];
            isFatefulEncounter = (pkx[0x1D] & 1) == 1;
            gender = (byte)((pkx[0x1D] >> 1) & 0x3);
            form = (byte)(pkx[0x1D] >> 3);
            evHp = pkx[0x1E];
            evAtk = pkx[0x1F];
            evDef = pkx[0x20];
            evSpAtk = pkx[0x22];
            evSpDef = pkx[0x23];
            evSpe = pkx[0x21];
            contestStatCool = pkx[0x24];
            contestStatBeauty = pkx[0x25];
            contestStatCute = pkx[0x26];
            contestStatSmart = pkx[0x27];
            contestStatTough = pkx[0x28];
            contestStatSheen = pkx[0x29];
            markings = pkx[0x2A];
            pkrsStrain = pkx[0x2B] >> 4;
            pkrsDuration = pkx[0x2B] % 0x10;

            // Block B
            nickname = Encoding.Unicode.GetString(pkx, 0x40, 24).TrimCString();
            // 0x58, 0x59 - unused
            move1 = BitConverter.ToUInt16(pkx, 0x5A);
            move2 = BitConverter.ToUInt16(pkx, 0x5C);
            move3 = BitConverter.ToUInt16(pkx, 0x5E);
            move4 = BitConverter.ToUInt16(pkx, 0x60);
            move1Pp = pkx[0x62];
            move2Pp = pkx[0x63];
            move3Pp = pkx[0x64];
            move4Pp = pkx[0x65];
            move1Ppu = pkx[0x66];
            move2Ppu = pkx[0x67];
            move3Ppu = pkx[0x68];
            move4Ppu = pkx[0x69];
            eggMove1 = BitConverter.ToUInt16(pkx, 0x6A);
            eggMove2 = BitConverter.ToUInt16(pkx, 0x6C);
            eggMove3 = BitConverter.ToUInt16(pkx, 0x6E);
            eggMove4 = BitConverter.ToUInt16(pkx, 0x70);

            // 0x72 - Super Training Flag - Passed with pkx to new form

            // 0x73 - unused/unknown
            uint IV32 = BitConverter.ToUInt32(pkx, 0x74);
            ivHp = IV32 & 0x1F;
            ivAtk = (IV32 >> 5) & 0x1F;
            ivDef = (IV32 >> 10) & 0x1F;
            ivSpe = (IV32 >> 15) & 0x1F;
            ivSpAtk = (IV32 >> 20) & 0x1F;
            ivSpDef = (IV32 >> 25) & 0x1F;
            isEgg = ((IV32 >> 30) & 1) != 0;
            isNick = ((IV32 >> 31)) != 0;

            // Block C
            notOT = Encoding.Unicode.GetString(pkx, 0x78, 24).TrimCString();
            bool notOTG = (pkx[0x92]) != 0;
            // Memory Editor edits everything else with pkx in a new form

            // Block D
            ot = Encoding.Unicode.GetString(pkx, 0xB0, 24).TrimCString();
            // 0xC8, 0xC9 - unused
            otFriendship = pkx[0xCA];
            otAffection = pkx[0xCB]; // Handled by Memory Editor
            // 0xCC, 0xCD, 0xCE, 0xCF, 0xD0
            eggDate = new Date(pkx[0xD1], pkx[0xD2], pkx[0xD3]);
            metDate = new Date(pkx[0xD4], pkx[0xD5], pkx[0xD6]);
            // 0xD7 - unused
            eggLocation = BitConverter.ToUInt16(pkx, 0xD8);
            metLocation = BitConverter.ToUInt16(pkx, 0xDA);
            ball = pkx[0xDC];
            levelMet = pkx[0xDD] & 0x7F;
            otGender = (pkx[0xDD]) >> 7;
            encounterType = pkx[0xDE];
            gamevers = pkx[0xDF];
            countryID = pkx[0xE0];
            regionID = pkx[0xE1];
            dsregID = pkx[0xE2];
            otLang = pkx[0xE3];

            hpType = (15 * ((ivHp & 1) + 2 * (ivAtk & 1) + 4 * (ivDef & 1) + 8 * (ivSpe & 1) + 16 * (ivSpAtk & 1) + 32 * (ivSpDef & 1))) / 63 + 1;

            tsv = (ushort)((tid ^ sid) >> 4);
            esv = (ushort)(((pid >> 16) ^ (pid & 0xFFFF)) >> 4);

            isShiny = (tsv == esv);
        }

        // Code by Kaphotics
        private static Uint8Array shuffleArray(Uint8Array pkx, uint sv)
        {
            Uint8Array ekx = new Uint8Array(pkx.length); Uint8ArrayHelper.Copy(pkx, 0, ekx, 0, 8);

            // Now to shuffle the blocks

            // Define Shuffle Order Structure
            var aloc = new byte[] { 0, 0, 0, 0, 0, 0, 1, 1, 2, 3, 2, 3, 1, 1, 2, 3, 2, 3, 1, 1, 2, 3, 2, 3 };
            var bloc = new byte[] { 1, 1, 2, 3, 2, 3, 0, 0, 0, 0, 0, 0, 2, 3, 1, 1, 3, 2, 2, 3, 1, 1, 3, 2 };
            var cloc = new byte[] { 2, 3, 1, 1, 3, 2, 2, 3, 1, 1, 3, 2, 0, 0, 0, 0, 0, 0, 3, 2, 3, 2, 1, 1 };
            var dloc = new byte[] { 3, 2, 3, 2, 1, 1, 3, 2, 3, 2, 1, 1, 3, 2, 3, 2, 1, 1, 0, 0, 0, 0, 0, 0 };

            // Get Shuffle Order
            var shlog = new byte[] { aloc[sv], bloc[sv], cloc[sv], dloc[sv] };

            // UnShuffle Away!
            for (int b = 0; b < 4; b++)
                Uint8ArrayHelper.Copy(pkx, 8 + 56 * shlog[b], ekx, 8 + 56 * b, 56);

            // Fill the Battle Stats back
            if (pkx.length > 232)
                Uint8ArrayHelper.Copy(pkx, 232, ekx, 232, 28);
            return ekx;
        }

        // Original code by Kaphotics
        public static Uint8Array decrypt(Uint8Array ekx)
        {
            Uint8Array pkx = new Uint8Array(232); Uint8ArrayHelper.Copy(ekx, 0, pkx, 0, 0xE8);
            uint pv = BitConverter.ToUInt32(pkx, 0);
            uint sv = (((pv & 0x3E000) >> 0xD) % 24);

            uint seed = pv;

            Uint16Array pkx16 = new Uint16Array(pkx.buffer);
            // Decrypt Blocks with RNG Seed
            for (int i = 4; i < 232/2; ++i)
            {
                seed = LCRNG.next(seed);
                pkx16[i] ^= (ushort)(seed >> 0x10);
            }

            // TODO: decrypt party stats?

            // Deshuffle
            pkx = shuffleArray(pkx, sv);
            return pkx;
        }

        // Original code by Kaphotics
        public static Uint8Array encrypt(Uint8Array pkx)
        {
            // Shuffle
            uint pv = BitConverter.ToUInt32(pkx, 0);
            uint sv = (((pv & 0x3E000) >> 0xD) % 24);

            Uint8Array ekx = new Uint8Array(pkx.length); Uint8ArrayHelper.Copy(pkx, 0, ekx, 0, pkx.length);

            // If I unshuffle 11 times, the 12th (decryption) will always decrypt to ABCD.
            // 2 x 3 x 4 = 12 (possible unshuffle loops -> total iterations)
            for (int i = 0; i < 11; i++)
                ekx = shuffleArray(ekx, sv);

            uint seed = pv;
            Uint16Array ekx16 = new Uint16Array(ekx.buffer);
            // Encrypt Blocks with RNG Seed
            for (int i = 4; i < 232/2; ++i)
            {
                seed = LCRNG.next(seed);
                ekx16[i] ^= (ushort)(seed >> 16);
            }

            // Encrypt the Party Stats
            seed = pv;
            if (pkx.length > 232)
                for (int i = 232/2; i < 260/2; ++i)
                {
                    seed = LCRNG.next(seed);
                    ekx16[i] ^= (ushort)(seed >> 16);
                }

            // Done
            return ekx;
        }

        // Code by Kaphotics
        public static bool verifyCHK(Uint8Array pkx)
        {
            // TODO refactor, probably pass Uint16Array
            ushort chk = 0;
            for (int i = 8; i < 232; i += 2) // Loop through the entire PKX
                chk += BitConverter.ToUInt16(pkx, i);

            ushort actualsum = BitConverter.ToUInt16(pkx, 0x6);
            if ((BitConverter.ToUInt16(pkx, 0x8) > 750) || (BitConverter.ToUInt16(pkx, 0x90) != 0))
                return false;
            return ((chk&0xFFFF) == actualsum);
        }

        public static byte getDloc(uint ec)
        {
            // Define Shuffle Order Structure
            var dloc = new byte[] { 3, 2, 3, 2, 1, 1, 3, 2, 3, 2, 1, 1, 3, 2, 3, 2, 1, 1, 0, 0, 0, 0, 0, 0 };
            uint sv = (((ec & 0x3E000) >> 0xD) % 24);

            return dloc[sv];
        }
    }
}

File: pkx.ts
"use strict";

import * as util from "./util";
import * as LCRNG from "./lcrng";

export default class Pkx {
    // Uints
    public ec: number;
    public pid: number;
    public exp: number;
    public evHp: number;
    public evAtk: number;
    public evDef: number;
    public evSpAtk: number;
    public evSpDef: number;
    public evSpe: number;
    public ivHp: number;
    public ivAtk: number;
    public ivDef: number;
    public ivSpe: number;
    public ivSpAtk: number;
    public ivSpDef: number;
    public contestStatCool: number;
    public contestStatBeauty: number;
    public contestStatCute: number;
    public contestStatSmart: number;
    public contestStatTough: number;
    public contestStatSheen: number;
    public markings: number;
    public hpType: number;

    // Ints
    public pkrsStrain: number;
    public pkrsDuration: number;
    public levelMet: number;
    public otGender: number;

    // Ushorts
    public ability: number;
    public abilityNum: number;
    public nature: number;
    public species: number;
    public heldItem: number;
    public tid: number;
    public sid: number;
    public tsv: number;
    public esv: number;
    public move1: number;
    public move2: number;
    public move3: number;
    public move4: number;
    public move1Pp: number;
    public move2Pp: number;
    public move3Pp: number;
    public move4Pp: number;
    public move1Ppu: number;
    public move2Ppu: number;
    public move3Ppu: number;
    public move4Ppu: number;
    public eggMove1: number;
    public eggMove2: number;
    public eggMove3: number;
    public eggMove4: number;
    public ribbonSet1: number;
    public ribbonSet2: number;
    public chk: number;
    public otFriendship: number;
    public otAffection: number;
    public eggLocation: number;
    public metLocation: number;
    public ball: number;
    public encounterType: number;
    public gameVersion: number;
    public countryID: number;
    public regionID: number;
    public dsregID: number;
    public otLang: number;

    // Shorts
    public box: number;
    public slot: number;

    // Bytes
    public ribbonSet3: number;
    public ribbonSet4: number;
    public form: number;
    public gender: number;

    // Longs
    public metDate: number;
    public eggDate: number;

    public isEgg: boolean;
    public isNick: boolean;
    public isShiny: boolean;
    public isGhost: boolean;
    public isFatefulEncounter: boolean;

    public data: number[];

    public nickname: string;
    public notOT: string;
    public ot: string;

    constructor(pkx: Uint8Array, box: number, slot: number, isGhost: boolean) {
        this.data = Array.from(pkx);
        this.box = box;
        this.slot = slot;
        this.isGhost = isGhost;

        var data: DataView = new DataView(pkx.buffer);

        this.ec = data.getUint32(0, true);
        this.chk = data.getUint16(6, true);
        this.species = data.getUint16(8, true);
        this.heldItem = data.getUint16(10, true);
        this.tid = data.getUint16(12, true);
        this.sid = data.getUint16(14, true);
        this.exp = data.getUint32(16, true);
        this.ability = pkx[20];
        this.abilityNum = pkx[21];
        // 0x16, 0x17 - unknown
        this.pid = data.getUint32(24, true);
        this.nature = pkx[28];
        this.isFatefulEncounter = (pkx[29] & 1) == 1;
        this.gender = (((pkx[29] >> 1) & 3) & 0xFF);
        this.form = ((pkx[29] >> 3) & 0xFF);
        this.evHp = pkx[30];
        this.evAtk = pkx[31];
        this.evDef = pkx[32];
        this.evSpAtk = pkx[34];
        this.evSpDef = pkx[35];
        this.evSpe = pkx[33];
        this.contestStatCool = pkx[36];
        this.contestStatBeauty = pkx[37];
        this.contestStatCute = pkx[38];
        this.contestStatSmart = pkx[39];
        this.contestStatTough = pkx[40];
        this.contestStatSheen = pkx[41];
        this.markings = pkx[42];
        this.pkrsStrain = pkx[43] >> 4;
        this.pkrsDuration = pkx[43] % 16;
        this.ribbonSet1 = data.getUint16(48, true);
        this.ribbonSet2 = data.getUint16(50, true);
        this.ribbonSet3 = pkx[52];
        this.ribbonSet4 = pkx[53];

        // Block B
        this.nickname = util.decodeUnicode16LE(pkx, 64, 24);
        // 0x58, 0x59 - unused
        this.move1 = data.getUint16(90, true);
        this.move2 = data.getUint16(92, true);
        this.move3 = data.getUint16(94, true);
        this.move4 = data.getUint16(96, true);
        this.move1Pp = pkx[98];
        this.move2Pp = pkx[99];
        this.move3Pp = pkx[100];
        this.move4Pp = pkx[101];
        this.move1Ppu = pkx[102];
        this.move2Ppu = pkx[103];
        this.move3Ppu = pkx[104];
        this.move4Ppu = pkx[105];
        this.eggMove1 = data.getUint16(106, true);
        this.eggMove2 = data.getUint16(108, true);
        this.eggMove3 = data.getUint16(110, true);
        this.eggMove4 = data.getUint16(112, true);

        // 0x72 - Super Training Flag - Passed with pkx to new form

        // 0x73 - unused/unknown
        var IV32 = data.getUint32(116, true);
        this.ivHp = IV32 & 31;
        this.ivAtk = (IV32 >> 5) & 31;
        this.ivDef = (IV32 >> 10) & 31;
        this.ivSpe = (IV32 >> 15) & 31;
        this.ivSpAtk = (IV32 >> 20) & 31;
        this.ivSpDef = (IV32 >> 25) & 31;
        this.isEgg = ((IV32 >> 30) & 1) != 0;
        this.isNick = (IV32 >> 31) != 0;

        // Block C
        this.notOT = util.decodeUnicode16LE(pkx, 120, 24);
        var notOTG = (pkx[146]) != 0;
        // Memory Editor edits everything else with pkx in a new form

        // Block D
        this.ot = util.decodeUnicode16LE(pkx, 176, 24);
        // 0xC8, 0xC9 - unused
        this.otFriendship = pkx[202];
        this.otAffection = pkx[203]; // Handled by Memory Editor
        // 0xCC, 0xCD, 0xCE, 0xCF, 0xD0
        this.eggDate = new Date(pkx[209] + 2000, pkx[210], pkx[211], 0, 0, 0, 0).getTime();
        this.metDate = new Date(pkx[212] + 2000, pkx[213], pkx[214], 0, 0, 0, 0).getTime();
        // 0xD7 - unused
        this.eggLocation = data.getUint16(216, true);
        this.metLocation = data.getUint16(218, true);
        this.ball = pkx[220];
        this.levelMet = pkx[221] & 127;
        this.otGender = (pkx[221]) >> 7;
        this.encounterType = pkx[222];
        this.gameVersion = pkx[223];
        this.countryID = pkx[224];
        this.regionID = pkx[225];
        this.dsregID = pkx[226];
        this.otLang = pkx[227];

        this.hpType = ((15 * ((this.ivHp & 1) + 2 * (this.ivAtk & 1) + 4 * (this.ivDef & 1) + 8 * (this.ivSpe & 1) + 16 * (this.ivSpAtk & 1) + 32 * (this.ivSpDef & 1))) / 63 | 0) + 1;

        this.tsv = (((this.tid ^ this.sid) >> 4) & 0xFFFF);
        this.esv = (((((this.pid >> 16) & 65535) ^ (this.pid & 65535)) >> 4) & 0xFFFF);

        this.isShiny = (this.tsv == this.esv);
    }

    static deshuffle(pkx: Uint8Array, sv: number) {
        var ekx = new Uint8Array(pkx.length);
        util.copy(pkx, 0, ekx, 0, 8);

        // Deshuffle order
        var sloc = [[ 0, 1, 2, 3 ],
                    [ 0, 1, 3, 2 ],
                    [ 0, 2, 1, 3 ],
                    [ 0, 3, 1, 2 ],
                    [ 0, 2, 3, 1 ],
                    [ 0, 3, 2, 1 ],
                    [ 1, 0, 2, 3 ],
                    [ 1, 0, 3, 2 ],
                    [ 2, 0, 1, 3 ],
                    [ 3, 0, 1, 2 ],
                    [ 2, 0, 3, 1 ],
                    [ 3, 0, 2, 1 ],
                    [ 1, 2, 0, 3 ],
                    [ 1, 3, 0, 2 ],
                    [ 2, 1, 0, 3 ],
                    [ 3, 1, 0, 2 ],
                    [ 2, 3, 0, 1 ],
                    [ 3, 2, 0, 1 ],
                    [ 1, 2, 3, 0 ],
                    [ 1, 3, 2, 0 ],
                    [ 2, 1, 3, 0 ],
                    [ 3, 1, 2, 0 ],
                    [ 2, 3, 1, 0 ],
                    [ 3, 2, 1, 0 ]];

        var shuffle = sloc[sv];

        for (var b = 0; b < 4; b++)
            util.copy(pkx, 8 + 56 * shuffle[b], ekx, 8 + 56 * b, 56);

        // Copy back party data
        if (pkx.length > 232)
            util.copy(pkx, 232, ekx, 232, 28);
        return ekx;
    }

    static shuffle(pkx: Uint8Array, sv: number) {
        var ekx = new Uint8Array(pkx.length);
        util.copy(pkx, 0, ekx, 0, 8);

        // Shuffle order
        var sloc = [[ 0, 1, 2, 3 ],
                    [ 0, 1, 3, 2 ],
                    [ 0, 2, 1, 3 ],
                    [ 0, 2, 3, 1 ],
                    [ 0, 3, 1, 2 ],
                    [ 0, 3, 2, 1 ],
                    [ 1, 0, 2, 3 ],
                    [ 1, 0, 3, 2 ],
                    [ 1, 2, 0, 3 ],
                    [ 1, 2, 3, 0 ],
                    [ 1, 3, 0, 2 ],
                    [ 1, 3, 2, 0 ],
                    [ 2, 0, 1, 3 ],
                    [ 2, 0, 3, 1 ],
                    [ 2, 1, 0, 3 ],
                    [ 2, 1, 3, 0 ],
                    [ 2, 3, 0, 1 ],
                    [ 2, 3, 1, 0 ],
                    [ 3, 0, 1, 2 ],
                    [ 3, 0, 2, 1 ],
                    [ 3, 1, 0, 2 ],
                    [ 3, 1, 2, 0 ],
                    [ 3, 2, 0, 1 ],
                    [ 3, 2, 1, 0 ]];

        var shuffle = sloc[sv];

        for (var b = 0; b < 4; b++)
            util.copy(pkx, 8 + 56 * shuffle[b], ekx, 8 + 56 * b, 56);

        // Copy back party data
        if (pkx.length > 232)
            util.copy(pkx, 232, ekx, 232, 28);
        return ekx;
    }

    static decrypt(ekx: Uint8Array) {
        var pkx = new Uint8Array(232);
        util.copy(ekx, 0, pkx, 0, 232);
        var pv = util.createDataView(pkx).getUint32(0, true);
        var sv = (((pv & 0x3E000) >> 0xD) % 24);
        var seed = pv;

        var pkx16 = util.createUint16Array(pkx);
        for (var i = 4; i < 232/2; ++i) {
            seed = LCRNG.next(seed);
            pkx16[i] ^= ((seed >> 0x10) & 0xFFFF);
        }

        // Decrypt party data
        seed = pv;
        if (pkx.length > 232) {
            for (var i = 232 / 2; i < 260 / 2; ++i) {
                seed = LCRNG.next(seed);
                pkx16[i] ^= ((seed >> 16) & 0xFFFF);
            }
        }

        pkx = Pkx.deshuffle(pkx, sv);
        return pkx;
    }

    static encrypt(pkx: Uint8Array) {
        var ekx = new Uint8Array(pkx.length);
        util.copy(pkx, 0, ekx, 0, 232);
        var pv = util.createDataView(pkx).getUint32(0, true);
        var sv = (((pv & 0x3E000) >> 0xD) % 24);

        ekx = Pkx.shuffle(ekx, sv);

        var seed = pv;
        var ekx16 = new Uint16Array(ekx.buffer);
        // Encrypt blocks with RNG generated key
        for (var i = 4; i < 232/2; ++i) {
            seed = LCRNG.next(seed);
            ekx16[i] ^= ((seed >> 16) & 0xFFFF);
        }

        // Encrypt the party data
        seed = pv;
        if (pkx.length > 232) {
            for (var i = 232 / 2; i < 260 / 2; ++i) {
                seed = LCRNG.next(seed);
                ekx16[i] ^= ((seed >> 16) & 0xFFFF);
            }
        }

        return ekx;
    }

    static calcChk(pkx: Uint8Array) {
        var chk = 0;
        var pkx16 = util.createUint16Array(pkx);
        for (var i = 8/2; i < 232/2; i++) {
            chk += pkx16[i];
        }
        return chk&0xFFFF;
    }

    static verifyChk(pkx: Uint8Array) {
        var chk = 0;
        var pkx16 = util.createUint16Array(pkx);
        for (var i = 8/2; i < 232/2; i++) {
            chk += pkx16[i];
        }

        var actualsum = pkx16[6/2];
        if (pkx16[8/2] > 750 || pkx16[144/2] != 0)
            return false;
        return (chk & 0xFFFF) == actualsum;
    }

    static getDloc(ec: number) {
        return [ 3, 2, 3, 2, 1, 1, 3, 2, 3, 2, 1, 1, 3, 2, 3, 2, 1, 1, 0, 0, 0, 0, 0, 0 ][
            ((ec & 0x3E000) >> 0xD) % 24
        ];
    }
}
